sdk/core/System.ClientModel/src/ModelReaderWriter/JsonPathReader.cs

```diff
@@ -0,0 +1,359 @@
+                Current = ReadNumber();
+                if (_consumed >= _length || _jsonPath[_consumed] != CloseBracket)
+                    throw new FormatException($"Invalid JsonPath syntax at position {_consumed}: expected ']' after number");
+
+                _consumed++; // Skip closing bracket
+                return true;
+
+            case SingleQuote or DoubleQuote:
+                if (Current.TokenType == JsonPathTokenType.PropertySeparator && _jsonPath[_consumed - 1] == Dot)
+                {
+                    Current = ReadProperty();
+                }
+                else
+                {
+                    Current = ReadQuotedString();
+                    if (_consumed >= _length || _jsonPath[_consumed] != CloseBracket)
+                        throw new FormatException($"Invalid JsonPath syntax at position {_consumed}: expected ']' after quoted string");
+
+                    _consumed++; // Skip closing bracket
+                }
+                return true;
+
+            default:
+                // Read a property name
+                Current = ReadProperty();
+                return true;
+        }
+    }
+
+    public JsonPathToken Peek()
+    {
+        if (Current.TokenType == JsonPathTokenType.End)
+            return Current;
+
+        if (_peeked.HasValue)
+            return _peeked.Token;
+
+        JsonPathReader local = this;
+        local.Read();
+        _peeked.HasValue = true;
+        _peeked.Consumed = local._consumed;
+        _peeked.Token = local.Current;
+
+        return _peeked.Token;
+    }
+
+    public ReadOnlySpan<byte> GetFirstProperty()
+    {
+        if (_jsonPath.IsRoot())
+            return _jsonPath;
+
+        if (!Read())
+            return ReadOnlySpan<byte>.Empty;
+
+        if (Current.TokenType != JsonPathTokenType.Root)
+            return ReadOnlySpan<byte>.Empty;
+
+        if (!Read())
+            return ReadOnlySpan<byte>.Empty;
+
+        if (Current.TokenType == JsonPathTokenType.PropertySeparator && !Read())
+            return ReadOnlySpan<byte>.Empty;
+
+        switch (Current.TokenType)
+        {
+            case JsonPathTokenType.Property:
+                return _jsonPath.Slice(0, _consumed);
+
+            case JsonPathTokenType.ArrayIndex:
+                return _jsonPath.Slice(0, _consumed);
+
+            default:
+                return ReadOnlySpan<byte>.Empty;
+        }
+    }
+
+    public ReadOnlySpan<byte> GetNextArray()
+    {
+        while (Read())
+        {
+            if (Current.TokenType == JsonPathTokenType.ArrayIndex)
+                return _jsonPath.Slice(0, _consumed);
+        }
+
+        return ReadOnlySpan<byte>.Empty;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private JsonPathToken ReadQuotedString()
+    {
+        var localJsonPath = _jsonPath;
+        byte quote = localJsonPath[_consumed];
+        int initial = ++_consumed; // Skip opening quote
```

Comment:
would it be useful to have Debug.Asserts for things like this, i.e. assert that it's indeed a quote.